File: artifacts/api-server/src/routes/grades.ts
import { Router, type IRouter } from "express";
import { eq, and } from "drizzle-orm";
import { db, gradesTable, studentsTable, assignmentsTable, coursesTable } from "@workspace/db";
import {
  ListGradesQueryParams,
  CreateGradeBody,
  UpdateGradeParams,
  UpdateGradeBody,
  DeleteGradeParams,
} from "@workspace/api-zod";
import { scoreToLetter, toPercent } from "../lib/gradeUtils";

const router: IRouter = Router();

router.get("/grades", async (req, res): Promise<void> => {
  const query = ListGradesQueryParams.safeParse(req.query);
  if (!query.success) {
    res.status(400).json({ error: query.error.message });
    return;
  }

  const { student_id, course_id, assignment_id } = query.data as { student_id?: number; course_id?: number; assignment_id?: number };
  const conditions = [];
  if (student_id !== undefined) conditions.push(eq(gradesTable.studentId, student_id));
  if (assignment_id !== undefined) conditions.push(eq(gradesTable.assignmentId, assignment_id));

  let rows;
  if (course_id !== undefined) {
    // Join through assignments to filter by course
    rows = await db
      .select({
        grade: gradesTable,
        studentName: studentsTable.name,
        assignmentName: assignmentsTable.name,
        maxScore: assignmentsTable.maxScore,
        courseId: assignmentsTable.courseId,
      })
      .from(gradesTable)
      .leftJoin(studentsTable, eq(gradesTable.studentId, studentsTable.id))
      .leftJoin(assignmentsTable, eq(gradesTable.assignmentId, assignmentsTable.id))
      .where(
        and(
          eq(assignmentsTable.courseId, course_id),
          ...(student_id !== undefined ? [eq(gradesTable.studentId, student_id)] : []),
          ...(assignment_id !== undefined ? [eq(gradesTable.assignmentId, assignment_id)] : []),
        ),
      )
      .orderBy(gradesTable.submittedAt);
  } else {
    rows = await db
      .select({
        grade: gradesTable,
        studentName: studentsTable.name,
        assignmentName: assignmentsTable.name,
        maxScore: assignmentsTable.maxScore,
        courseId: assignmentsTable.courseId,
      })
      .from(gradesTable)
      .leftJoin(studentsTable, eq(gradesTable.studentId, studentsTable.id))
      .leftJoin(assignmentsTable, eq(gradesTable.assignmentId, assignmentsTable.id))
      .where(conditions.length > 0 ? and(...conditions) : undefined)
      .orderBy(gradesTable.submittedAt);
  }

  res.json(
    rows.map((r) => {
      const pct = r.maxScore ? toPercent(Number(r.grade.score), Number(r.maxScore)) : null;
      return {
        ...r.grade,
        student_id: r.grade.studentId,
        assignment_id: r.grade.assignmentId,
        score: Number(r.grade.score),
        percentage: pct !== null ? Math.round(pct * 10) / 10 : null,
        letter_grade: pct !== null ? scoreToLetter(pct) : null,
        submitted_at: r.grade.submittedAt.toISOString(),
        student_name: r.studentName ?? null,
        assignment_name: r.assignmentName ?? null,
        course_id: r.courseId ?? null,
      };
    }),
  );
});

router.post("/grades", async (req, res): Promise<void> => {
  const parsed = CreateGradeBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const d = parsed.data as { student_id: number; assignment_id: number; score: number; feedback?: string };

  const [grade] = await db
    .insert(gradesTable)
    .values({
      studentId: d.student_id,
      assignmentId: d.assignment_id,
      score: String(d.score),
      feedback: d.feedback ?? null,
    })
    .returning();

  const asgn = await db.select().from(assignmentsTable).where(eq(assignmentsTable.id, d.assignment_id)).limit(1);
  const pct = asgn[0] ? toPercent(Number(grade.score), Number(asgn[0].maxScore)) : null;

  res.status(201).json({
    ...grade,
    student_id: grade.studentId,
    assignment_id: grade.assignmentId,
    score: Number(grade.score),
    percentage: pct !== null ? Math.round(pct * 10) / 10 : null,
    letter_grade: pct !== null ? scoreToLetter(pct) : null,
    submitted_at: grade.submittedAt.toISOString(),
    student_name: null,
    assignment_name: null,
    course_id: asgn[0]?.courseId ?? null,
  });
});

router.put("/grades/:id", async (req, res): Promise<void> => {
  const params = UpdateGradeParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  const parsed = UpdateGradeBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const d = parsed.data as { score?: number; feedback?: string };
  const updates: Record<string, unknown> = {};
  if (d.score !== undefined) updates.score = String(d.score);
  if (d.feedback !== undefined) updates.feedback = d.feedback;

  const [grade] = await db.update(gradesTable).set(updates).where(eq(gradesTable.id, params.data.id)).returning();
  if (!grade) {
    res.status(404).json({ error: "Grade not found" });
    return;
  }

  const asgn = await db.select().from(assignmentsTable).where(eq(assignmentsTable.id, grade.assignmentId)).limit(1);
  const pct = asgn[0] ? toPercent(Number(grade.score), Number(asgn[0].maxScore)) : null;

  res.json({
    ...grade,
    student_id: grade.studentId,
    assignment_id: grade.assignmentId,
    score: Number(grade.score),
    percentage: pct !== null ? Math.round(pct * 10) / 10 : null,
    letter_grade: pct !== null ? scoreToLetter(pct) : null,
    submitted_at: grade.submittedAt.toISOString(),
    student_name: null,
    assignment_name: null,
    course_id: asgn[0]?.courseId ?? null,
  });
});

router.delete("/grades/:id", async (req, res): Promise<void> => {
  const params = DeleteGradeParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }
  await db.delete(gradesTable).where(eq(gradesTable.id, params.data.id));
  res.sendStatus(204);
});

export default router;

File: artifacts/server/src/routes/grades.py
"""Grade routes — full CRUD."""
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from src.db.session import get_db
from src.db.models import GradeRow, AssignmentRow, StudentRow
from src.domain.grade_utils import to_percent, score_to_letter

router = APIRouter()


def _fmt(g: GradeRow, asgn: Optional[AssignmentRow] = None, student_name=None):
    pct = to_percent(float(g.score), float(asgn.max_score)) if asgn else None
    return {
        "id": g.id,
        "student_id": g.student_id,
        "assignment_id": g.assignment_id,
        "score": float(g.score),
        "feedback": g.feedback,
        "submitted_at": g.submitted_at.isoformat(),
        "percentage": round(pct * 10) / 10 if pct is not None else None,
        "letter_grade": score_to_letter(pct) if pct is not None else None,
        "student_name": student_name,
        "assignment_name": asgn.name if asgn else None,
        "course_id": asgn.course_id if asgn else None,
    }


class CreateGradeBody(BaseModel):
    student_id: int
    assignment_id: int
    score: float
    feedback: Optional[str] = None


class UpdateGradeBody(BaseModel):
    score: Optional[float] = None
    feedback: Optional[str] = None


@router.get("/grades")
def list_grades(
    student_id: Optional[int] = Query(None),
    course_id: Optional[int] = Query(None),
    assignment_id: Optional[int] = Query(None),
    db: Session = Depends(get_db),
):
    q = (
        db.query(GradeRow, AssignmentRow, StudentRow.name)
        .outerjoin(AssignmentRow, GradeRow.assignment_id == AssignmentRow.id)
        .outerjoin(StudentRow, GradeRow.student_id == StudentRow.id)
    )
    if student_id is not None:
        q = q.filter(GradeRow.student_id == student_id)
    if assignment_id is not None:
        q = q.filter(GradeRow.assignment_id == assignment_id)
    if course_id is not None:
        q = q.filter(AssignmentRow.course_id == course_id)
    rows = q.order_by(GradeRow.submitted_at).all()
    return [_fmt(g, a, sname) for g, a, sname in rows]


@router.post("/grades", status_code=201)
def create_grade(body: CreateGradeBody, db: Session = Depends(get_db)):
    row = GradeRow(
        student_id=body.student_id,
        assignment_id=body.assignment_id,
        score=str(body.score),
        feedback=body.feedback,
    )
    db.add(row)
    db.commit()
    db.refresh(row)
    asgn = db.query(AssignmentRow).filter(AssignmentRow.id == body.assignment_id).first()
    return _fmt(row, asgn)


@router.put("/grades/upsert")
def upsert_grade(body: CreateGradeBody, db: Session = Depends(get_db)):
    """Create or update a grade for a student+assignment pair."""
    existing = db.query(GradeRow).filter(
        GradeRow.student_id == body.student_id,
        GradeRow.assignment_id == body.assignment_id,
    ).first()
    if existing:
        existing.score = str(body.score)
        if body.feedback is not None:
            existing.feedback = body.feedback
        db.commit()
        db.refresh(existing)
        asgn = db.query(AssignmentRow).filter(AssignmentRow.id == existing.assignment_id).first()
        return _fmt(existing, asgn)
    else:
        row = GradeRow(
            student_id=body.student_id,
            assignment_id=body.assignment_id,
            score=str(body.score),
            feedback=body.feedback,
        )
        db.add(row)
        db.commit()
        db.refresh(row)
        asgn = db.query(AssignmentRow).filter(AssignmentRow.id == body.assignment_id).first()
        return _fmt(row, asgn)


@router.put("/grades/{grade_id}")
def update_grade(grade_id: int, body: UpdateGradeBody, db: Session = Depends(get_db)):
    g = db.query(GradeRow).filter(GradeRow.id == grade_id).first()
    if not g:
        raise HTTPException(status_code=404, detail="Grade not found")
    if body.score is not None:
        g.score = str(body.score)
    if body.feedback is not None:
        g.feedback = body.feedback
    db.commit()
    db.refresh(g)
    asgn = db.query(AssignmentRow).filter(AssignmentRow.id == g.assignment_id).first()
    return _fmt(g, asgn)


@router.delete("/grades/{grade_id}", status_code=204)
def delete_grade(grade_id: int, db: Session = Depends(get_db)):
    g = db.query(GradeRow).filter(GradeRow.id == grade_id).first()
    if g:
        db.delete(g)
        db.commit()
